### PR TITLE
Write a systemd service file when using systemd

### PR DESCRIPTION
**Pull request by petems**
* Works for any system with systemd service provider
* Adds spec for new case also

New PR over https://github.com/voxpupuli/puppet-dhcp/pull/129/files

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -151,31 +151,44 @@
...
-      }
+  if $::service_provider == 'systemd' {
+    file { '/etc/systemd/system/dhcpd.service':
+      ensure  => present,
+      owner   => 'root',
+      group   => 'root',
+      mode    => '0644',
+      before  => Package[$packagename],
+      notify  => Service[$servicename],
+      content => template('dhcp/dhcpd.service'),
```
Comment (bastelfreak) [inline manifests/init.pp]: you can use epp if you like that more

File: manifests/init.pp [hunk 2]
```diff
@@ -151,31 +151,44 @@
...
-      file{ '/etc/default/isc-dhcp-server':
-        ensure  => present,
-        owner   => 'root',
-        group   => 'root',
-        mode    => '0644',
-        before  => Package[$packagename],
-        notify  => $service_notify_real,
-        content => template('dhcp/debian/default_isc-dhcp-server'),
-      }
+  if $::service_provider == 'systemd' {
```
Comment (bastelfreak) [inline manifests/init.pp]: could you use the new facts hash here?

File: manifests/init.pp [hunk 3]
```diff
@@ -151,31 +151,44 @@
...
-        owner   => 'root',
-        group   => 'root',
-        mode    => '0644',
-        before  => Package[$packagename],
-        notify  => $service_notify_real,
-        content => template('dhcp/redhat/sysconfig-dhcpd'),
+
+  } else {
+    # Only debian and ubuntu have this style of defaults for startup.
+    case $::osfamily {
```
Comment (bastelfreak) [inline manifests/init.pp]: Could you use the new facts hash here?

File: manifests/init.pp (hunk 1 above)
Comment (bastelfreak) [reply manifests/init.pp]: We need a notify here for systemctl reload. The resource should come from https://github.com/camptocamp/puppet-systemd#overview. Could you please replace the file resource with ::systemd::unit_file ?

File: manifests/init.pp (hunk 2 above)
Comment (petems) [reply manifests/init.pp]: Fixed 👍

File: manifests/init.pp (hunk 3 above)
Comment (petems) [reply manifests/init.pp]: Fixed 👍

**Comment by ekohl** [comment]
I'd still prefer a drop in just for EL7 rather than fully managing the file for all systemd platforms. Debian Strech still uses a init.d file with overrides in ```/etc/default``` and I don't know how well it goes when you move from init.d to service. Upgrades might break.

File: templates/dhcpd.service [hunk 4]
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=DHCPv4 Server Daemon
+Documentation=man:dhcpd(8) man:dhcpd.conf(5)
+Wants=network-online.target
+After=network-online.target
+After=time-sync.target
+
+[Service]
+Type=notify
+ExecStart=/usr/sbin/dhcpd -f -cf /etc/dhcp/dhcpd.conf -user dhcpd -group dhcpd --no-pid <%= Array(@dhcp_interfaces).join(' ') %>
```
Comment (bastelfreak) [inline templates/dhcpd.service]: $dhcp_dir should be used instead of hardcoding the path to the config file. At least on arch it is /etc/dhcpd.conf

Comment (petems) [reply templates/dhcpd.service]: Good point, and adding in Arch spec also 👍

**Comment by juniorsysadmin** [comment]
> I'd still prefer a drop in just for EL7
I agree.

**Comment by bastelfreak** [comment]
> > I'd still prefer a drop in just for EL7

>I agree.

can we then get archlinux and centos please. Also ping @petems :)

**Comment by petems** [comment]
@bastelfreak Done! Simplified to just EL-7 and Archlinux 👍

File: manifests/init.pp [hunk 5]
```diff
@@ -158,31 +158,60 @@
...
-        ensure  => present,
-        owner   => 'root',
-        group   => 'root',
-        mode    => '0644',
-        before  => Package[$packagename],
-        notify  => $service_notify_real,
-        content => template('dhcp/debian/default_isc-dhcp-server'),
+
+
+  case $facts['osfamily'] {
```
Comment (bastelfreak) [inline manifests/init.pp]: facts hash \o/

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -22,7 +22,8 @@
 RSpec.configure do |c|
   default_facts = {
     puppetversion: Puppet.version,
-    facterversion: Facter.version
+    facterversion: Facter.version,
+    service_provider: 'none'
```
Comment (bastelfreak) [inline spec/spec_helper.rb]: this is managed by our modulesync_config. Can you please put it into spec/default_module_facts.yml?

Also: Why none? Is this a value that any system reports? I expected systemd or sysv or something similar.

Comment (petems) [reply spec/spec_helper.rb]: Fixed by removing it 👍

File: manifests/init.pp (hunk 5 above)
Comment (petems) [reply manifests/init.pp]: 🎆

**Pull request merged**